packages/doenetml-worker-rust/lib-doenetml-core/src/core/dast/flat_dast/untagged_flat_dast_compactify.rs

```diff
@@ -0,0 +1,184 @@
+                                _ => None,
+                            })
+                        })
+                        .collect()
+                } else {
+                    vec![]
+                }
+            }
+            _ => vec![],
+        }
+    }
+
+    /// Given `ref_index_map`, which is a vector of old indices to new indices,
+    /// rewrite all `Ref` nodes to point to the new indices.
+    fn shift_references(&mut self, ref_index_map: &[Index]) {
+        match self {
+            FlatNode::Element(elm) => {
+                for child in elm.children.iter_mut() {
+                    if let UntaggedContent::Ref(idx) = child {
+                        *idx = ref_index_map[*idx];
+                    }
+                }
+                for attr in elm.attributes.iter_mut() {
+                    for child in attr.children.iter_mut() {
+                        if let UntaggedContent::Ref(idx) = child {
+                            *idx = ref_index_map[*idx];
+                        }
+                    }
+                }
+                if let Some(parent) = &mut elm.parent {
+                    *parent = ref_index_map[*parent];
+                }
+            }
+            FlatNode::FunctionMacro(function_macro) => {
+                if let Some(inputs) = &mut function_macro.input {
+                    for inputs in inputs.iter_mut() {
+                        for child in inputs.iter_mut() {
+                            if let UntaggedContent::Ref(idx) = child {
+                                *idx = ref_index_map[*idx];
+                            }
+                        }
+                    }
+                }
+                if let Some(parent) = &mut function_macro.parent {
+                    *parent = ref_index_map[*parent];
+                }
+            }
+            FlatNode::Macro(macro_) => {
+                if let Some(parent) = &mut macro_.parent {
+                    *parent = ref_index_map[*parent];
+                }
+            }
+            FlatNode::Error(err) => {
+                if let Some(parent) = &mut err.parent {
+                    *parent = ref_index_map[*parent];
+                }
+            }
+        }
+    }
+}
+
+impl FlatRoot {
+    /// Remove any unreferenced nodes and shrink the `nodes` array to fit.
+    /// Indices are adjusted to reflect the new positions of the nodes.
+    pub fn compactify(&mut self) {
+        let mut is_referenced: Vec<bool> =
+            std::iter::repeat(false).take(self.nodes.len()).collect();
+        let mut to_visit = self
+            .children
+            .iter()
+            .filter_map(|node| match node {
+                UntaggedContent::Ref(idx) => Some(*idx),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        while let Some(idx) = to_visit.pop() {
+            // Don't visit a node's children more than once.
+            if is_referenced[idx] {
+                continue;
```

Comment:
What would cause this algorithm to visit a node more than once? Are you attempting to be able to handle loops in the graph? It might be better to panic in that situation.

At any rate, if I put a panic in here instead, the tests still pass. So if there is a reason for this test, it's not covered by the tests.

Reply:
Yes. The idea was to prevent loops from breaking the code. I would rather have this code as loop-tolerant even if it's functionality we don't need.

Reply:
Sounds good. Maybe add a comment to this effect?